contracts/p1/BackingManager.sol

```diff
@@ -146,14 +130,106 @@ contract BackingManagerP1 is TradingP1, IBackingManager {
                     if (req.sellAmount > bal) stRSR.seizeRSR(req.sellAmount - bal);
                 }
 
-                tryTrade(req);
+                openTrade(req);
             } else {
                 // Haircut time
                 compromiseBasketsNeeded(basketsHeld.bottom, basketsNeeded);
             }
         }
     }
 
+    /// Maintain the overall backing policy in an atomic swap via a dutch auction
+    /// @dev Caller must have granted tokenIn allowances for required tokenIn bal
+    /// @param tokenIn The ERC20 token provided by the caller
+    /// @param tokenOut The ERC20 token being purchased by the caller
+    /// @param amountOut {qTokenOut} The exact quantity of tokenOut being purchased
+    /// @return s The exact Swap performed
+    /// @custom:interaction RCEI
+    function swap(
+        IERC20 tokenIn,
+        IERC20 tokenOut,
+        uint256 amountOut
+    ) external returns (Swap memory s) {
+        // == Refresh ==
+        assetRegistry.refresh();
+        furnace.melt();
+
+        requireCanTrade(0);
+
+        // executeSwap if storage auction already exists
+        if (tradeEnd > block.timestamp) {
+            return executeSwap(dutchAuctions[tradeEnd], amountOut);
+        }
+
+        // === Checks/Effects ===
+
+        require(tradeEnd + dutchAuctionLength > block.timestamp, "no dutch auction ongoing");
+        tradeEnd += dutchAuctionLength;
+
+        // Same TradeRequest from manageTokens()
+        (bool doTrade, TradeRequest memory req) = RecollateralizationLibP1
+            .prepareRecollateralizationTrade(this, basketHandler.basketsHeldBy(address(this)));
+
+        require(doTrade, "swap not available");
+        require(tokenIn == req.buy.erc20() && tokenOut == req.sell.erc20(), "ERC20 swap expired");
```

Comment:
not sure this is the right error message? "Swap tokens mismatch" might make more sense?

Reply:
I could go either way here. I thought the mismatch error message might not be super clear for a searcher. The BackingManager is highly opinionated about what swap it offers, and all it takes is for some minute appreciation to work out differently for the currently offered swap to change. My thinking was that this kind of error message was more clear, but I don't feel strongly.